src/allmydata/test/test_magic_folder.py

```diff
@@ -1518,6 +1525,64 @@ def test_delete(self):
         self.assertTrue(node is not None, "Failed to find %r in DMD" % (path,))
         self.failUnlessEqual(metadata['version'], 1)
 
+    @defer.inlineCallbacks
+    def test_batched_process(self):
+        # setup: get at least two items into the deque
+        path0 = os.path.join(self.local_dir, u'foo')
+        yield self.fileops.write(path0, 'foo\n')
+        path1 = os.path.join(self.local_dir, u'bar')
+        yield self.fileops.write(path1, 'bar\n')
+
+        # get the status before we've processed anything
+        upstatus0 = list(self.magicfolder.uploader.get_status())
+        upstatus1 = []
+
+        def one_item(item):
+            # grab status after we've processed a single item
+            us = list(self.magicfolder.uploader.get_status())
+            upstatus1.extend(us)
+        one_d = self.magicfolder.uploader.set_hook('item_processed')
+        # can't 'yield' here because the hook isn't called until
+        # inside iterate()
+        one_d.addCallbacks(one_item, self.fail)
+
+        yield iterate_uploader(self.magicfolder)
+        yield iterate_uploader(self.magicfolder)  # req'd for windows; not sure why?
+
+        # status we got each time should be the same
+        self.assertEqual(len(upstatus0), len(upstatus1))
+        for item0, item1 in zip(upstatus0, upstatus1):
+            self.assertEqual(item0, item1)
+
+    @defer.inlineCallbacks
+    def test_real_notify_failure(self):
+
+        def bad_stuff(*args, **kw):
+            # the function we replaced would trigger this hook, so we
+            # have to or things will stall
```

Comment:
Another possibility is to actually call the function that's being replaced.

Questionable whether this is an improvement over what's here now... Either way private implementation details are being touched.  Maybe the suggestion is an improvement becanse then it's just one implementation detail being touched twice (patched and called by the patched version) instead of (patch `_real_notify` and call `_call_hook` from the patched version).

Reply:
okay, this sounds good .. I agree also about the internals, but also the whole thing we're testing there is "internal" so meh